### PR TITLE
TBD: Shadow DOM streamer

### PR DESCRIPTION
**Pull request by dvoytenko**
"Since Shadow DOM streaming is not coming to us, ..."

This exact utility can be reused for streaming friendly iframes as well.

### CODE REVIEW

File: src/utils/dom-stream.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+    let next = null;
+    let depth = 0;
+    if (this.inputPoint_.nextSibling) {
+      // Simplest situation: the next sibling has already been parsed.
+      next = this.inputPoint_.nextSibling;
+    } else {
+      // Maybe this is a last child? Try parents.
+      let localDepth = 0;
+      for (let n = this.inputPoint_.parentNode;
+              n && n != this.input_;
```
Comment (cramforce) [inline src/utils/dom-stream.js]: nit: indent looks off

File: src/utils/dom-stream.js [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    // There are two phases:
+    // 1. "advance". This phase waits for a sibling node to the pending node
+    //    or a singling node to one of its ancestors. As soon as sibling node
+    //    is found, the parse of pending node is complete and it can be merged
+    //    into the target DOM.
+    // 2. "split". If no sibling nodes are found, this phase finds the nearest
+    //    descendant of this node with a sibling. If found, it imports the
+    //    partial DOM path and the found descendant node into the target DOM.
+    //    There are limitations of how deep the input DOM can be split and some
+    //    nodes are not allowed to be split, such as "table" and "ul".
```
Comment (cramforce) [inline src/utils/dom-stream.js]: Would you mind adding HTML examples to this comment as a sort of timeline for when things happen?

File: src/utils/dom-stream.js [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+  }
+
+  /** @private */
+  split_() {
+    // Find a node that can be split.
+    let node = this.inputPoint_;
+    let next = null;
+    let depth = 0;
+    while (!next &&
+            node.firstChild &&
```
Comment (cramforce) [inline src/utils/dom-stream.js]: nit: indent

File: src/utils/dom-stream.js (hunk 2 above)
Comment (cramforce) [inline src/utils/dom-stream.js]: I guess the tests kind of do this already, but might be nice here.

**Comment by dvoytenko** [comment]
All, please hold on on this, thanks to @jakearchibald and https://twitter.com/jaffathecake/status/865585897172946944 this can be made quite a bit simpler.

**Pull request closed without merging**